flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java

```diff
@@ -0,0 +1,379 @@
+						}
+					}
+				}
+			}
+		}
+		finally {
+			// this signals the consumer thread that no more work is to be done
+			consumerThread.shutdown();
+		}
+
+		// on a clean exit, wait for the runner thread
+		try {
+			consumerThread.join();
+		}
+		catch (InterruptedException e) {
+			// may be the result of a wake-up interruption after an exception.
+			// we ignore this here and only restore the interruption state
+			Thread.currentThread().interrupt();
+		}
+	}
+
+	@Override
+	public void cancel() {
+		// flag the main thread to exit. A thread interrupt will come anyways.
+		running = false;
+		handover.close();
+		consumerThread.shutdown();
+	}
+
+	@Override
+	protected TopicPartition createKafkaPartitionHandle(KafkaTopicPartition partition) {
+		return new TopicPartition(partition.getTopic(), partition.getPartition());
+	}
+
+	@Override
+	protected void doCommitInternalOffsetsToKafka(
+			Map<KafkaTopicPartition, Long> offsets,
+			@Nonnull KafkaCommitCallback commitCallback) throws Exception {
+		@SuppressWarnings("unchecked")
+		List<KafkaTopicPartitionState<TopicPartition>> partitions = subscribedPartitionStates();
+
+		Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = new HashMap<>(partitions.size());
+
+		for (KafkaTopicPartitionState<TopicPartition> partition : partitions) {
+			Long lastProcessedOffset = offsets.get(partition.getKafkaTopicPartition());
+			if (lastProcessedOffset != null) {
+				checkState(lastProcessedOffset >= 0, "Illegal offset value to commit");
+
+				// committed offsets through the KafkaConsumer need to be 1 more than the last processed offset.
+				// This does not affect Flink's checkpoints/saved state.
+				long offsetToCommit = lastProcessedOffset + 1;
+
+				offsetsToCommit.put(partition.getKafkaPartitionHandle(), new OffsetAndMetadata(offsetToCommit));
+				partition.setCommittedOffset(offsetToCommit);
+			}
+		}
+
+		// record the work to be committed by the main consumer thread and make sure the consumer notices that
+		consumerThread.setOffsetsToCommit(offsetsToCommit, commitCallback);
+	}
+
+	private String getFetcherName() {
+		return "Kafka Shuffle Fetcher";
+	}
+
+	/**
+	 * An element in a KafkaShuffle. Can be a record or a Watermark.
+	 */
+	@VisibleForTesting
+	public abstract static class KafkaShuffleElement<T> {
+
+		public boolean isRecord() {
+			return getClass() == KafkaShuffleRecord.class;
+		}
+
+		public boolean isWatermark() {
+			return getClass() == KafkaShuffleWatermark.class;
+		}
+
+		public KafkaShuffleRecord<T> asRecord() {
+			return (KafkaShuffleRecord<T>) this;
+		}
+
+		public KafkaShuffleWatermark asWatermark() {
+			return (KafkaShuffleWatermark) this;
+		}
+	}
+
+	/**
+	 * A watermark element in a KafkaShuffle. It includes
+	 * - subtask index where the watermark is coming from
+	 * - watermark timestamp
+	 */
+	@VisibleForTesting
+	public static class KafkaShuffleWatermark<T> extends KafkaShuffleElement<T> {
```

Comment:
`KafkaShuffleWatermark` is always used as a raw type, so please get rid of the type parameter. 

I also think that `KafkaShuffleElement<T>` should be without type parameter and only `asRecord` has it.
```
	public <T> KafkaShuffleRecord<T> asRecord() {
			return (KafkaShuffleRecord<T>) this;
	}
```

Reply:
Yes, good catch.